### ISSUE TITLE
exec: AND expressions don't short-circuit

### ISSUE THREAD

**Issue by solongordon** [cockroachdb/cockroach#40688]
@rohany pointed out that ANDs can short-circuit if the left side is false. The vectorized engine does not do this.
```
root@127.0.0.1:53988/defaultdb> create table t (a bool, b int);
CREATE TABLE

root@127.0.0.1:53988/defaultdb> insert into t values (false, 0);
INSERT 1

root@127.0.0.1:53988/defaultdb> set vectorize=off;
SET

root@127.0.0.1:53988/defaultdb> select a and 1/b = 1 from t;
  ?column?
+----------+
   false
(1 row)

root@127.0.0.1:53988/defaultdb> set vectorize=experimental_always;
SET

root@127.0.0.1:53988/defaultdb> select a and 1/b = 1 from t;
pq: division by zero
```
According to Rohan the SQL standard doesn't mandate this short-circuiting behavior, but both Postgres and distsql do implement it.

The easiest way to fix this (though probably not the most efficient) would be to do the same thing we did with ORs and rewrite the AND as a CASE expression. (`a OR b` becomes `CASE WHEN a THEN b ELSE false`.)

**Comment by rohany** [comment]
Looking around it seems like the order of evaluation of a && b is undefined? We might have just gotten lucky here.

**Comment by yuzefovich** [comment]
Not knowing about this issue I included this problem in #40913 which will be fixed #40847, so I'm closing this issue as a duplicate.
